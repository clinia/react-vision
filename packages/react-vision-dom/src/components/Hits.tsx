import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { createClassNames } from '../core/utils';
import { translatable } from 'react-vision-core';
import { defaultLoadingIndicator } from './defaultComponents';

export type Record = {
  id: string;
};

type HitProps = {
  record: Record;
};

type Props = {
  records: Record[];
  loading?: boolean;
  className?: string;
  style?: React.CSSProperties;
  noResultsFound?: React.ReactNode;
  hit?: React.FunctionComponent<HitProps>;
  translate?: any;
};

const DefaultHit: React.FC<HitProps> = ({ record }) => {
  return (
    <div
      style={{
        borderBottom: '1px solid #bbb',
        paddingBottom: '5px',
        marginBottom: '5px',
        wordBreak: 'break-all',
      }}
    >
      <h4>{record.id}</h4>
      <code>{JSON.stringify(record, null, 2)}</code>
    </div>
  );
};

DefaultHit.propTypes = {
  record: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
};

const cx = createClassNames('hits');

const Hits: React.FunctionComponent<Props> = ({
  records,
  className = '',
  style = {},
  hit: HitComponent = DefaultHit,
  noResultsFound,
  translate,
  loading,
}) => {
  return loading ? (
    defaultLoadingIndicator(cx)
  ) : (
    <div className={classnames(cx(''), className)} style={style}>
      <div className={cx('list')}>
        {Array.isArray(records) && records.length > 0
          ? records.map(record => (
              <div className={cx('item')} key={record.id}>
                <HitComponent record={record} />
              </div>
            ))
          : noResultsFound || (
              <div className={cx('empty')}>{translate('emptySearch')}</div>
            )}
      </div>
    </div>
  );
};

const HitPropTypes = PropTypes.shape({
  id: PropTypes.string.isRequired,
});

Hits.propTypes = {
  records: PropTypes.arrayOf(HitPropTypes.isRequired).isRequired,
  className: PropTypes.string,
  noResultsFound: PropTypes.func,
  translate: PropTypes.func,
  loading: PropTypes.bool,
  style: PropTypes.object,
  hit: PropTypes.func,
};

export default translatable({
  emptySearch: 'No results were found',
})(Hits);
