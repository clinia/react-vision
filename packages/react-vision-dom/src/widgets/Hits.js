import { connectHits } from 'react-vision-core';
import Hits from '../components/Hits';

export default connectHits(Hits);
