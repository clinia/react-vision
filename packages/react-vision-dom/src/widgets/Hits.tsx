import { connectHits } from 'react-vision-core';
import Hits from '../components/Hits';

/**
 * Displays a list of hits.
 *
 * @name Hits
 * @kind widget
 * @propType {Component} [hitComponent] - Component used for rendering each hit from
 *   the results. If it is not provided the rendering defaults to displaying the
 *   hit in its JSON form. The component will be called with a `record` prop.
 * @propType {node} [noResultsFound] - Change the apparence of the default empty results found.
 * @propType {string} [className] - Add a custom CSS class to the component
 * @propType {object} [style] - Add a custom React.CSSProperties object to the Hits container
 * @themeKey cvi-Hits - the root div of the widget
 * @themeKey cvi-Hits-list - the list of results
 * @themeKey cvi-Hits-item - the hit list item
 * @translationkey emptySearch - The text shown on empty results
 * @example
 * import React from 'react';
 * import cliniasearch from 'cliniasearch/lite';
 * import { Vision, Hits } from 'react-vision-dom';
 *
 * const searchClient = cliniasearch(
 *   'TODO',
 *   'test'
 * );
 * const App = () => (
 *   <Vision
 *     searchClient={searchClient}
 *     indexName="health_facility"
 *   >
 *     <Hits />
 *   </Vision>
 * );
 */

export default connectHits(Hits);
