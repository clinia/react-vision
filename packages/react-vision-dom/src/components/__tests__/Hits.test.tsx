import React from 'react';
import renderer from 'react-test-renderer';
import Hits from '../Hits';

describe('Hits', () => {
  type Props = { record: any };

  const Hit = ({ record }: Props) => <div id={record.id} />;

  it('accepts a hit prop', () => {
    const records = [{ id: '0' }, { id: '1' }, { id: '2' }];

    const tree = renderer.create(<Hits hit={Hit} records={records} />);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('accepts a custom className', () => {
    const records = [{ id: '0' }, { id: '1' }, { id: '2' }];

    const tree = renderer.create(
      <Hits className="MyCustomHits" hit={Hit} records={records} />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
