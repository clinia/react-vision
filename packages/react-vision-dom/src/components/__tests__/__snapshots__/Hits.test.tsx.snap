// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hits accepts a custom className 1`] = `
<div
  className="cvi-Hits MyCustomHits"
  style={Object {}}
>
  <div
    className="cvi-Hits-list"
  >
    <div
      className="cvi-Hits-item"
    >
      <div
        id="0"
      />
    </div>
    <div
      className="cvi-Hits-item"
    >
      <div
        id="1"
      />
    </div>
    <div
      className="cvi-Hits-item"
    >
      <div
        id="2"
      />
    </div>
  </div>
</div>
`;

exports[`Hits accepts a hit prop 1`] = `
<div
  className="cvi-Hits"
  style={Object {}}
>
  <div
    className="cvi-Hits-list"
  >
    <div
      className="cvi-Hits-item"
    >
      <div
        id="0"
      />
    </div>
    <div
      className="cvi-Hits-item"
    >
      <div
        id="1"
      />
    </div>
    <div
      className="cvi-Hits-item"
    >
      <div
        id="2"
      />
    </div>
  </div>
</div>
`;
